File: code/backend/python/app/adapters/rest/match/inventory_controller.py
"""Steps 34 & 35 — FastAPI controller for the inventory and the resources.

  GET  /api/gameplay/{uuid_match}/inventory            -> 200 | 401 | 404
  POST /api/gameplay/{uuid_match}/inventory/use-item   -> 200 | 400 | 401 | 404 | 409
  POST /api/gameplay/{uuid_match}/inventory/drop-item  -> 200 | 400 | 401 | 404 | 409
  GET  /api/gameplay/{uuid_match}/resources            -> 200 | 401 | 404

use-item answers with the execute-event payload, because an item carrying a SADNESS effect
can trigger the step-30 overflow or coma and the frontend then reuses its event handler
almost unchanged. On an item usage `eventUuid` and `eventType` are null and `card` is the
item's own card.

Both request bodies name `itemInstanceUuid`: the uuid of the INVENTORY ROW
(`items[].uuid`), never the story item's `items[].itemUuid`.
"""
import time
from typing import Any, Dict, List

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from app.adapters.rest.match.event_controller import _result_to_camel
from app.core.models.match.match_models import ItemInstanceInfo
from app.core.ports.match.inventory_ports import InventoryError, InventoryPort

# Not-found for what does not exist; conflict for a state the player could act on.
_STATUS_BY_CODE = {
    InventoryError.MATCH_NOT_FOUND: 404,
    InventoryError.ITEM_NOT_FOUND: 404,
    InventoryError.MATCH_NOT_RUNNING: 409,
    InventoryError.SLEEPING: 409,
    InventoryError.COMA: 409,
    InventoryError.ITEM_NOT_CONSUMABLE: 409,
    InventoryError.ITEM_CLASS_NOT_PERMITTED: 409,
    InventoryError.ITEM_CLASS_PROHIBITED: 409,
}


def _error(code: str, message: str, http_status: int) -> JSONResponse:
    return JSONResponse(
        status_code=http_status,
        content={"error": code, "message": message, "timestamp": int(time.time() * 1000)},
    )


def item_to_camel(i: ItemInstanceInfo) -> Dict[str, Any]:
    """The one item shape, shared with the match /info players[] projection."""
    return {
        "uuid": i.uuid,
        "itemUuid": i.item_uuid,
        "name": i.name,
        "weight": i.weight,
        "amount": i.amount,
        "state": i.state,
        "idCard": i.id_card,
        "card": i.card,
        "isConsumabile": i.is_consumabile,
    }


def _inventory_to_camel(v: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "matchUuid": v["match_uuid"],
        "characterUuid": v["character_uuid"],
        "items": [item_to_camel(i) for i in v["items"]],
        "weight": v["weight"],
        "weightMax": v["weight_max"],
    }


def _drop_to_camel(v: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "matchUuid": v["match_uuid"],
        "characterUuid": v["character_uuid"],
        "itemInstanceUuid": v["item_instance_uuid"],
        "itemUuid": v["item_uuid"],
        "amountDropped": v["amount_dropped"],
        "weight": v["weight"],
        "weightMax": v["weight_max"],
        # Always true — the inventory and the carried weight both changed.
        "refreshRecommended": True,
    }


def _resources_to_camel(v: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "matchUuid": v["match_uuid"],
        "characterUuid": v["character_uuid"],
        "food": v["food"],
        "magic": v["magic"],
        # Singular, as everywhere in the API. react-game maps it to `coins`.
        "coin": v["coin"],
        "weight": v["weight"],
        "weightMax": v["weight_max"],
    }


class InventoryController:
    def __init__(self, inventory_port: InventoryPort):
        self.inventory_port = inventory_port
        self.router = APIRouter()
        self.router.add_api_route(
            "/api/gameplay/{uuid_match}/inventory", self.inventory, methods=["GET"])
        self.router.add_api_route(
            "/api/gameplay/{uuid_match}/inventory/use-item", self.use_item, methods=["POST"])
        self.router.add_api_route(
            "/api/gameplay/{uuid_match}/inventory/drop-item", self.drop_item, methods=["POST"])
        self.router.add_api_route(
            "/api/gameplay/{uuid_match}/resources", self.resources, methods=["GET"])

    async def inventory(self, uuid_match: str, request: Request, lang: str = "en"):
        user_uuid = getattr(request.state, "user_uuid", None)
        if not user_uuid:
            return _error("UNAUTHENTICATED", "User identity is missing", 401)
        try:
            view = self.inventory_port.list_inventory(uuid_match, user_uuid, lang)
        except InventoryError as exc:
            return _error(exc.code, exc.message, _STATUS_BY_CODE.get(exc.code, 409))
        return JSONResponse(status_code=200, content=_inventory_to_camel(view))

    async def use_item(self, uuid_match: str, request: Request, lang: str = "en"):
        user_uuid = getattr(request.state, "user_uuid", None)
        if not user_uuid:
            return _error("UNAUTHENTICATED", "User identity is missing", 401)
        item_uuid = await self._item_instance_uuid(request)
        if not item_uuid:
            return _error("MISSING_ITEM", "itemInstanceUuid is required", 400)
        try:
            result = self.inventory_port.use_item(uuid_match, user_uuid, item_uuid, lang)
        except InventoryError as exc:
            return _error(exc.code, exc.message, _STATUS_BY_CODE.get(exc.code, 409))
        return JSONResponse(status_code=200, content=_result_to_camel(result))

    async def drop_item(self, uuid_match: str, request: Request):
        user_uuid = getattr(request.state, "user_uuid", None)
        if not user_uuid:
            return _error("UNAUTHENTICATED", "User identity is missing", 401)
        item_uuid = await self._item_instance_uuid(request)
        if not item_uuid:
            return _error("MISSING_ITEM", "itemInstanceUuid is required", 400)
        try:
            view = self.inventory_port.drop_item(uuid_match, user_uuid, item_uuid)
        except InventoryError as exc:
            return _error(exc.code, exc.message, _STATUS_BY_CODE.get(exc.code, 409))
        return JSONResponse(status_code=200, content=_drop_to_camel(view))

    async def resources(self, uuid_match: str, request: Request):
        user_uuid = getattr(request.state, "user_uuid", None)
        if not user_uuid:
            return _error("UNAUTHENTICATED", "User identity is missing", 401)
        try:
            view = self.inventory_port.get_resources(uuid_match, user_uuid)
        except InventoryError as exc:
            return _error(exc.code, exc.message, _STATUS_BY_CODE.get(exc.code, 409))
        return JSONResponse(status_code=200, content=_resources_to_camel(view))

    @staticmethod
    async def _item_instance_uuid(request: Request):
        try:
            body = await request.json()
        except Exception:
            body = {}
        value = (body or {}).get("itemInstanceUuid")
        return value if value and str(value).strip() else None

File: code/backend/java/adapter-rest/src/main/java/games/paths/adapters/rest/controller/match/InventoryController.java
package games.paths.adapters.rest.controller.match;

import games.paths.adapters.rest.dto.DropItemRequest;
import games.paths.adapters.rest.dto.DropItemResponse;
import games.paths.adapters.rest.dto.ExecuteEventResponse;
import games.paths.adapters.rest.dto.InventoryResponse;
import games.paths.adapters.rest.dto.ResourcesResponse;
import games.paths.adapters.rest.dto.UseItemRequest;
import games.paths.core.port.match.InventoryPort;
import games.paths.core.port.match.InventoryPort.InventoryException;

import jakarta.servlet.http.HttpServletRequest;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * InventoryController - REST adapter for the inventory (Step 34) and the resources
 * (Step 35) of the calling character.
 *
 * <ul>
 *   <li>GET  /api/gameplay/&#123;uuidMatch&#125;/inventory — what the caller carries</li>
 *   <li>POST /api/gameplay/&#123;uuidMatch&#125;/inventory/use-item — consume one item</li>
 *   <li>POST /api/gameplay/&#123;uuidMatch&#125;/inventory/drop-item — discard one item</li>
 *   <li>GET  /api/gameplay/&#123;uuidMatch&#125;/resources — food, magic, coin, weight</li>
 * </ul>
 *
 * <p>Use-item answers with {@link ExecuteEventResponse}, the execute-event payload,
 * because an item carrying a SADNESS effect can trigger the Step 30 overflow or coma —
 * the frontend then reuses its event handler almost unchanged. On that response an item
 * usage reports {@code eventUuid: null} and {@code eventType: null}, and {@code card} is
 * the item's own card.</p>
 *
 * <p>Both request bodies name {@code itemInstanceUuid}: the uuid of the INVENTORY ROW
 * ({@code items[].uuid}), never the story item's {@code items[].itemUuid}.</p>
 */
@RestController
public class InventoryController {

    private final InventoryPort inventoryPort;

    public InventoryController(InventoryPort inventoryPort) {
        this.inventoryPort = inventoryPort;
    }

    @GetMapping("/api/gameplay/{uuidMatch}/inventory")
    public ResponseEntity<Object> inventory(@PathVariable String uuidMatch,
                                            @RequestParam(required = false) String lang,
                                            HttpServletRequest request) {
        String userUuid = userUuid(request);
        if (userUuid == null) {
            return error(HttpStatus.UNAUTHORIZED, "UNAUTHENTICATED", "User identity is missing");
        }
        try {
            return ResponseEntity.ok(InventoryResponse.fromModel(
                    inventoryPort.listInventory(uuidMatch, userUuid, lang)));
        } catch (InventoryException ex) {
            return error(mapStatus(ex.getCode()), ex.getCode().name(), ex.getMessage());
        }
    }

    @PostMapping("/api/gameplay/{uuidMatch}/inventory/use-item")
    public ResponseEntity<Object> useItem(@PathVariable String uuidMatch,
                                          @RequestParam(required = false) String lang,
                                          @RequestBody(required = false) UseItemRequest body,
                                          HttpServletRequest request) {
        String userUuid = userUuid(request);
        if (userUuid == null) {
            return error(HttpStatus.UNAUTHORIZED, "UNAUTHENTICATED", "User identity is missing");
        }
        String itemInstanceUuid = body == null ? null : body.getItemInstanceUuid();
        if (itemInstanceUuid == null || itemInstanceUuid.isBlank()) {
            return error(HttpStatus.BAD_REQUEST, "MISSING_ITEM", "itemInstanceUuid is required");
        }
        try {
            return ResponseEntity.ok(ExecuteEventResponse.fromModel(
                    inventoryPort.useItem(uuidMatch, userUuid, itemInstanceUuid, lang)));
        } catch (InventoryException ex) {
            return error(mapStatus(ex.getCode()), ex.getCode().name(), ex.getMessage());
        }
    }

    @PostMapping("/api/gameplay/{uuidMatch}/inventory/drop-item")
    public ResponseEntity<Object> dropItem(@PathVariable String uuidMatch,
                                           @RequestBody(required = false) DropItemRequest body,
                                           HttpServletRequest request) {
        String userUuid = userUuid(request);
        if (userUuid == null) {
            return error(HttpStatus.UNAUTHORIZED, "UNAUTHENTICATED", "User identity is missing");
        }
        String itemInstanceUuid = body == null ? null : body.getItemInstanceUuid();
        if (itemInstanceUuid == null || itemInstanceUuid.isBlank()) {
            return error(HttpStatus.BAD_REQUEST, "MISSING_ITEM", "itemInstanceUuid is required");
        }
        try {
            return ResponseEntity.ok(DropItemResponse.fromModel(
                    inventoryPort.dropItem(uuidMatch, userUuid, itemInstanceUuid)));
        } catch (InventoryException ex) {
            return error(mapStatus(ex.getCode()), ex.getCode().name(), ex.getMessage());
        }
    }

    @GetMapping("/api/gameplay/{uuidMatch}/resources")
    public ResponseEntity<Object> resources(@PathVariable String uuidMatch,
                                            HttpServletRequest request) {
        String userUuid = userUuid(request);
        if (userUuid == null) {
            return error(HttpStatus.UNAUTHORIZED, "UNAUTHENTICATED", "User identity is missing");
        }
        try {
            return ResponseEntity.ok(ResourcesResponse.fromModel(
                    inventoryPort.getResources(uuidMatch, userUuid)));
        } catch (InventoryException ex) {
            return error(mapStatus(ex.getCode()), ex.getCode().name(), ex.getMessage());
        }
    }

    private static String userUuid(HttpServletRequest request) {
        String u = (String) request.getAttribute("userUuid");
        return (u == null || u.isBlank()) ? null : u;
    }

    private static ResponseEntity<Object> error(HttpStatus status, String code, String message) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("error", code);
        body.put("message", message);
        body.put("timestamp", System.currentTimeMillis());
        return ResponseEntity.status(status).body(body);
    }

    /** Not-found for a missing entity; conflict for a state the player could act on. */
    private static HttpStatus mapStatus(InventoryException.Code code) {
        return switch (code) {
            case MATCH_NOT_FOUND, ITEM_NOT_FOUND -> HttpStatus.NOT_FOUND;
            case MATCH_NOT_RUNNING, SLEEPING, COMA, ITEM_NOT_CONSUMABLE,
                 ITEM_CLASS_NOT_PERMITTED, ITEM_CLASS_PROHIBITED -> HttpStatus.CONFLICT;
        };
    }
}
